### PR TITLE
Fix vocab_size inconsistency for sampler

### PR DESCRIPTION
**Pull request by tongyx361**
Fixing #340

### CODE REVIEW

File: vllm/engine/llm_engine.py [hunk 1]
```diff
@@ -86,18 +86,20 @@ def __init__(
...
-
         self.tokenizer = get_tokenizer(
             model_config.tokenizer,
             tokenizer_mode=model_config.tokenizer_mode,
             trust_remote_code=model_config.trust_remote_code,
             tokenizer_revision=model_config.tokenizer_revision,
             revision=model_config.revision)
+        model_config.hf_config.sampler_vocab_size = min(
+            len(self.tokenizer), model_config.hf_config.vocab_size)
+        self.cache_config = cache_config
```
Comment (simon-mo) [inline vllm/engine/llm_engine.py:97]: These lines don't need to move right?

Comment (tongyx361) [reply vllm/engine/llm_engine.py:97]: Yes, `model_config.hf_config.sampler_vocab_size = min(len(self.tokenizer), model_config.hf_config.vocab_size)` is for `sampler = Sampler(config..sampler_vocab_size)`.

**Comment by zhuohan123** [comment]
@tongyx361 What exact bug is this PR fixing?

**Comment by tongyx361** [comment]
> @tongyx361 What exact bug is this PR fixing?

c.f. https://github.com/vllm-project/vllm/issues/340#issuecomment-1653326230

```
Traceback (most recent call last):
  File "vllm-none-problem-repro.py", line 21, in <module>
    out = llm.generate(input, sampling_params)
  File "/llm-bench/vllm-src/vllm/entrypoints/llm.py", line 127, in generate
    return self._run_engine(use_tqdm)
  File "/llm-bench/vllm-src/vllm/entrypoints/llm.py", line 147, in _run_engine
    step_outputs = self.llm_engine.step()
  File "/llm-bench/vllm-src/vllm/engine/llm_engine.py", line 246, in step
    self._decode_sequences(seq_groups)
  File "/llm-bench/vllm-src/vllm/engine/llm_engine.py", line 263, in _decode_sequences
    new_token, new_output_text = detokenize_incrementally(
  File "/llm-bench/vllm-src/vllm/transformers_utils/tokenizer.py", line 73, in detokenize_incrementally
    output_text = tokenizer.convert_tokens_to_string(output_tokens)
  File "/opt/conda/lib/python3.8/site-packages/transformers/tokenization_utils_fast.py", line 533, in convert_tokens_to_string
    return self.backend_tokenizer.decoder.decode(tokens)
TypeError: argument 'tokens': 'NoneType' object cannot be converted to 'PyString
```

> The reason is that for some models there can be a mismatch between the `config.vocab_size` and the `len(tokenizer)`. The model outputs a distribution over tokens in the range vocab_size, but only tokens in the range len(tokenizer) should actually be sampled. The remaining tokens are just padding and when sampling these tokens and decoding them, the result will be None instead of a string and so the exception will be thrown.

**Comment by creatorrr** [comment]
@zhuohan123 @simon-mo does this PR need more work? We are facing this issue in production

**Comment by creatorrr** [comment]
I see that the added `lora` padding complicates this PR quite a bit

**Comment by GennVa** [comment]
@zhuohan123 Are there any updates on a solution for this issue? Thanks

**Comment by creatorrr** [comment]
I ended up adding extra tokens to the tokenizer to make it a multiple of 32 and it works

**Comment by zhuohan123** [comment]
Sorry for the delay. This PR is a little bit stalled. I will try to fix this in another PR with some other changes in the `Sampler`

**Comment by GennVa** [comment]
@zhuohan123 Okay thanks! What is the tracking number of this PR#?

**Comment by GennVa** [comment]
@creatorrr Okay thanks, I have in my case a qwen tokenizer of 151648 (151643 + 5 added tokens) that is a multiple of 32, a vocab.json size of 151643 and a vocab_size in config.json of 151936. Tried to add the 5 added tokens in vocab.json and in tokenizer vocab but still same error.

**Pull request closed without merging**